rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -688,8 +789,17 @@ extern "C" rmw_node_t * rmw_create_node(
 #endif
 
   dds_qos_t * qos = dds_create_qos();
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    security_options, "rmw_create_node: Unable to create qos", return nullptr);
   std::string user_data = get_node_user_data(name, namespace_);
   dds_qset_userdata(qos, user_data.c_str(), user_data.size());
+
+  if (security_options->enforce_security) {
+    if (configure_qos_for_security(qos, security_options) != RMW_RET_OK) {
+      return nullptr;
```

Comment:
This, too, can leak the domain reference, and moreover leaks ``qos``. Isn't C just a lovely language, always being on the lookout for ways to get you? 😄

Reply:
> Isn't C just a lovely language, always being on the lookout for ways to get you?

Regarding that: is there a reason you're not making more substantial use of C++ features, here?

Reply:
Not really, the RMW layer is intended to be in C++. Part of the reason that it is so C-like is that I am much more experienced with C than with C++ and so whenever I try to write C++ it still ends up looking like C. There is also the fact that Cyclone is in C, that this is therefore relying on a C API, and that some of the niceties of C++ (like destructors) are simply not available on these objects.

There is an implementation of a standard C++11 API of DDS waiting for approval by the Eclipse Foundation's IP review process. That interface covers all the standard features but not all the additional tricks. At some point it might be worth rewriting the RMW implementation to use that API, it's only about a handful of lines of code anyway.

Reply:
> Part of the reason that it is so C-like is that I am much more experienced with C than with C++ and so whenever I try to write C++ it still ends up looking like C.

Fair enough, that's a perfectly valid reason.

> There is also the fact that Cyclone is in C, that this is therefore relying on a C API, and that some of the niceties of C++ (like destructors) are simply not available on these objects.

In case you're unaware, shared/unique ptrs can be handed custom deleters that might help with that particular issue.

Reply:
> In case you're unaware, shared/unique ptrs can be handed custom deleters that might help with that particular issue.

That sounds like something worth reading up on, thanks!

Reply:
@eboasson better yet, define a destructors for your classes. Then you don't need to pass a custom deleter to unique_ptr/shared_ptr

Reply:
@rotu that's not always possible:

> [...] this is therefore relying on a C API, and that some of the niceties of C++ (like destructors) are simply not available on these objects.

Reply:
I find it's easier to maintain a wrapper class than a deleter, but YMMV. Where there's a matched initializer + deleter pair, I find it easiest to make them into a constructor + destructor. With a factory method + deleter pair, i prefer to use a class method + protected constructor + destructor.

Reply:
@eboasson, this should all be covered with 1ca2269.